compiler/src/dotty/tools/dotc/transform/Pickler.scala

```diff
@@ -33,30 +43,160 @@ object Pickler {
    */
   inline val ParallelPickling = true
 
-  class EarlyFileWriter private (writer: TastyWriter, origin: AbstractFile):
-    def this(dest: AbstractFile)(using @constructorOnly ctx: Context) = this(TastyWriter(dest), dest)
+  /**A holder for syncronization points and reports when writing TASTy asynchronously.
+   * The callbacks should only be called once.
+   */
+  class AsyncTastyHolder private (
+      val earlyOut: AbstractFile, incCallback: IncrementalCallback | Null)(using @constructorOnly ex: ExecutionContext):
+    import scala.concurrent.Future as StdFuture
+    import scala.concurrent.Await
+    import scala.concurrent.duration.Duration
+    import AsyncTastyHolder.Signal
+
+    private val _cancel = AtomicBoolean(false)
+
+    /**Cancel any outstanding work.
+     * This should be done at the end of a run, e.g. if there were errors that prevented reaching the backend. */
+    def cancel(): Unit =
+      while
+        val cancelled = _cancel.get()
+        !cancelled && !_cancel.compareAndSet(false, true)
+      do ()
```

Comment:
The only state transition here is `false -> true`. Moreover, in case of double-`cancel()`, it's much better to ignore than to hang. So the canonical way to cancel with an atomic boolean would instead be:
```suggestion
      if _cancel.compareAndSet(false, true) then
        performCancellationOperations() // they trySuccess calls
      else
        // someone else already cancelled, so nothing to do
        ()
```

Reply:
Thanks! I have also adjusted other usages of atomic reference/boolean